waggle-dance-core/src/main/java/com/hotels/bdp/waggledance/mapping/model/MetaStoreMappingFactoryImpl.java

```diff
@@ -63,4 +73,34 @@ public String prefixNameFor(AbstractMetaStore federatedMetaStore) {
     return prefixNamingStrategy.apply(federatedMetaStore);
   }
 
+  private CloseableThriftHiveMetastoreIface newUnreachableMetatstoreClient(AbstractMetaStore metaStore) {
+    return (CloseableThriftHiveMetastoreIface) Proxy.newProxyInstance(getClass().getClassLoader(),
+        new Class[] { CloseableThriftHiveMetastoreIface.class },
+        new UnreachableMetastoreClientInvocationHandler(metaStore.getName()));
+  }
+
+  /**
+   * Handler that refuses to be open and will throw exceptions for any of the methods, serves as a dummy client if the
+   * real one can't be created due to connection (i.e. tunneling) issues.
+   */
+  private static class UnreachableMetastoreClientInvocationHandler implements InvocationHandler {
+
+    private final String name;
+
+    private UnreachableMetastoreClientInvocationHandler(String name) {
+      this.name = name;
+    }
+
+    @Override
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+      switch (method.getName()) {
+      case "isOpen":
+        return false;
```

Comment:
Does this mean later on WD will think this thing is closed and try create a new one and attempt a reconnect? I'm just trying to think about the sequence of events here - while the remote is unreachable how often will it check that it's back and then recover?

Reply:
User(hive/spark client) opens a session does a request for the duration of that session this client lives. So if you do a query (show databases) it won't show databases for the broken metastore. If you do a query on the specific database in that metastore (prefix and all) it will fail like hive would fail if you try to query a database that doesn't exist. 
If you close your client and reconnect it will check all metastores again.

Reply:
OK, makes sense, so for a call to a Thrift method it will create one of these, the specific method call will throw the below exception which WD will log and ignore and this will carry on until you reconnect at which point it will attempt to connect again etc.

Reply:
yes, you have to reconnect. If you are doing multiple queries inside the hive-client it will use the disconnected one.